### DISCUSSION TITLE
Berkeley to SQLite?

### DISCUSSION THREAD

**Discussion by qlpqlp** [dogecoin/dogecoin#2764] [Q&A] [answered]
Hello to all :)

Is there a time that you are thinking on migrating to SQLite ?

Berkeley gives some headaches wen growing bigger :)

and ofcourse, its a lot more easy for devs to explore and do external queries :)

Thanks to all in advance.

**Comment by vertiond** [comment, marked as answer]
This is the proposed timeline upstream: https://github.com/bitcoin/bitcoin/issues/20160

Note that Dogecoin is currently working on merging in upstream 0.21.

**Comment by patricklodder** [comment]
> Berkeley gives some headaches wen growing bigger :)

bdb is only used for the wallet though. For which use-case is your wallet so big that you outgrow bdb?

**Reply by qlpqlp** [reply]
Thats the point :)

If an online store wants to create one Dogecoin Address to each cliente and it grows to 1M clients, could it handle?
or, if the online store wants to create an Dogecoin Address to each product? exemple Alixpress, Ebay, Amazon, that has millions of products, can the core wallet handle?

**Reply by patricklodder** [reply]
No. You shouldn't use the Dogecoin Core wallet for that for security reasons: if you have a hot online wallet that does millions of transactions, the value held would benefit from industrial-grade hardware security to protect the keys. 

It's also feature poor when looking at enterprise needs. Custom indexing is not a luxury because you want to make decisions on transaction characteristics (eg if a tx sets an RBF flag, you may want to change the required number of confirmations and if your tx gets stuck, you may want to automatically bump/rebroadcast)

Long story short, Dogecoin Core is not an enterprise software, it's a reference client. In an enterprise setting I would run:

1. A public cluster of them with the wallet and all optional indexing disabled, and large, permissive mempools to get the most diverse set of peers possible.
2. A couple of private nodes, also without wallet, that only connect into the cluster that replicate transactions and blocks into an enterprise grade message queue, like Apache Kafka. May want to enable txindex on these (or a separate set) for handling consumer queries. 
3. Queue consumers that persist into a (commercial, relational) database, consumers that implement risk assessment protocols, consumers that trigger consolidation scripts, and so forth
4. At least 2 HSMs that securely store all key material
5. Isolated hosts with microservices that have access rights to request the HSMs to sign particular transactions.

and so on.

**Reply by qlpqlp** [reply]
Much wow!
Such enterprise advice!

Alredy testing since Saturday descentralized Arweave DB indexing options in conjunction with storj  that i have already tested last month and have an live test here: 
http://descentralized.what-is-dogecoin.com/

Objective, build a secure 💯 descentralized anda scalable environment for Dogecoin node :)
